File: server/models/deworming.py
import uuid
from datetime import datetime, date, timezone

from sqlalchemy import String, DateTime, Date, ForeignKey, Boolean
from sqlalchemy.orm import Mapped, mapped_column, relationship

from server.database.base import Base


def _uuid() -> str:
    return str(uuid.uuid4())


def _utcnow() -> datetime:
    return datetime.now(timezone.utc)


class Deworming(Base):
    __tablename__ = "deworming_records"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=_uuid)
    pet_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("pets.id", ondelete="CASCADE"), nullable=False, index=True
    )
    owner_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )

    product_name: Mapped[str] = mapped_column(String(120), nullable=False)
    date_given: Mapped[date] = mapped_column(Date, nullable=False)
    next_due_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    dosage: Mapped[str | None] = mapped_column(String(80), nullable=True)
    administered_by: Mapped[str | None] = mapped_column(String(120), nullable=True)
    notes: Mapped[str | None] = mapped_column(String(1000), nullable=True)
    reminder_enabled: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=_utcnow, nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=_utcnow, onupdate=_utcnow, nullable=False
    )

    pet = relationship("Pet", backref="deworming_records")

File: client/types/deworming.ts
export interface Deworming {
  id: string;
  petId: string;
  ownerId: string;
  productName: string;
  dateGiven: string;
  nextDueDate: string | null;
  dosage: string | null;
  administeredBy: string | null;
  notes: string | null;
  reminderEnabled: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface DewormingFormPayload {
  petId: string;
  productName: string;
  dateGiven: string;
  nextDueDate?: string;
  dosage?: string;
  administeredBy?: string;
  notes?: string;
  reminderEnabled: boolean;
}

interface RawDeworming {
  id: string;
  pet_id: string;
  owner_id: string;
  product_name: string;
  date_given: string;
  next_due_date: string | null;
  dosage: string | null;
  administered_by: string | null;
  notes: string | null;
  reminder_enabled: boolean;
  created_at: string;
  updated_at: string;
}

export function mapDeworming(raw: RawDeworming): Deworming {
  return {
    id: raw.id,
    petId: raw.pet_id,
    ownerId: raw.owner_id,
    productName: raw.product_name,
    dateGiven: raw.date_given,
    nextDueDate: raw.next_due_date,
    dosage: raw.dosage,
    administeredBy: raw.administered_by,
    notes: raw.notes,
    reminderEnabled: raw.reminder_enabled,
    createdAt: raw.created_at,
    updatedAt: raw.updated_at,
  };
}

export function toDewormingPayload(payload: DewormingFormPayload) {
  return {
    pet_id: payload.petId,
    product_name: payload.productName,
    date_given: payload.dateGiven,
    next_due_date: payload.nextDueDate || null,
    dosage: payload.dosage || null,
    administered_by: payload.administeredBy || null,
    notes: payload.notes || null,
    reminder_enabled: payload.reminderEnabled,
  };
}
